target_iceberg/utils.py

```diff
@@ -114,5 +114,33 @@ def deduplicate_table(table: pa.Table) -> pa.Table:
     return table.combine_chunks().group_by(table.column_names).aggregate([])
 
 
-def schemas_match(schema1: pa.Schema, schema2: pa.Schema) -> bool:
-    return schema1.equals(schema2, check_metadata=False)
+def are_types_equivalent(type1: pa.DataType, type2: pa.DataType) -> bool:
+    # We treat string and large_string as same type
+    return ({type1, type2} <= {pa.string(), pa.large_string()}
+            or type1 == type2)
+
+
+def schemas_match(old: pa.Schema, new: pa.Schema) -> bool:
+    old_fields = {field.name: field for field in old}
+    new_fields = {field.name: field for field in new}
+    all_field_names = set(old_fields.keys()) | set(new_fields.keys())
+
+    differences = []
+
+    for name in sorted(all_field_names):
+        f_old = old_fields.get(name)
+        f_new = new_fields.get(name)
+
+        if f_old and not f_new:
+            differences.append(f"Field '{name}' only in old schema: {f_old}")
+        elif not f_old and f_new:
+            differences.append(f"Field '{name}' only in new schema: {f_new}")
+        elif not are_types_equivalent(f_old.type, f_new.type):
+            differences.append(f"Field '{name}' type differs:\n  old: {f_old}\n  new: {f_new}")
+        elif f_old.nullable != f_new.nullable:
```

Comment:
Do we care about nullability?

Reply:
Good question. I believe in theory someone can change a non nullable field to a nullable field and then that is an actual differences in schema, which warrants table recreation?

Reply:
In spark/ganymedes, we are setting all fields in the Iceberg table as `nullable` irrespective of what's in avdl as nullability is checked by `safeAs` method. Not sure how we are ensuring nullability here.

Iceberg allows going from non-nullable to nullable field with a `ALTER TABLE` statement, so we can also  allow schema evolution without table recreation. Not sure if we should handle this case here or not.

Reply:
hmm, not sure as well, we could add an explicit check to run an alter table, but I'm not sure if that's really needed, given that the primary use case for this is writing google sheet data and I don't think recreating the table if nullability changes is an issue. I could also add code which makes everything nullable prior to writing to iceberg, but I'm not sure if it's worth it. Personally I would leave this logic as is, as it makes sense and is consistent and if there are any issues we can change how we handle nullability here.

Reply:
Sound good.